File: travel-agent-strands/src/agents/sites_agent.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""Sites specialist tool — port of agents/sites-agent.ts.

Exports a single ``query_sites`` Strands tool. The function-level docstring
becomes the tool description; argument docstrings become parameter descriptions.
"""

from __future__ import annotations

from typing import List, Optional

from strands import tool

from ..data.attractions import attractions
from ..data.types import Attraction


@tool
def query_sites(query: str, day_of_week: Optional[str] = None) -> str:
    """Query information about attractions in the city including opening hours, closure days, visit duration, ticket prices, and booking requirements. Use this to look up attraction details before planning routes.

    Args:
        query: Search query for attractions (name, category, or 'all').
        day_of_week: Day to check availability (e.g., 'Monday'). When set,
            only attractions open on that day are returned.

    Returns:
        Formatted multi-line text describing the matching attractions.
    """
    filtered: List[Attraction]

    if query.lower() == "all":
        filtered = list(attractions)
    else:
        lower_query = query.lower()
        filtered = [
            a
            for a in attractions
            if lower_query in a.name.lower() or a.category.lower() == lower_query
        ]

    # Filter by day availability if specified
    if day_of_week:
        filtered = [a for a in filtered if day_of_week not in a.closure_days]

    # Build text content with booking advisories
    lines: List[str] = []

    if not filtered:
        lines.append(f'No attractions found matching "{query}".')
    else:
        for attraction in filtered:
            lines.append(f"**{attraction.name}** ({attraction.category})")
            lines.append(f"  {attraction.description}")
            lines.append(f"Hours: {attraction.open_time}–{attraction.close_time}")
            if attraction.closure_days:
                lines.append(f"Closed: {', '.join(attraction.closure_days)}")
            lines.append(
                f"Visit duration: {attraction.visit_duration_minutes} minutes"
            )
            lines.append(f"Ticket price: ${attraction.ticket_price}")

            if attraction.advance_booking_required:
                lines.append(
                    f"  ⚠️ BOOKING ADVISORY: Advance booking is required for "
                    f"{attraction.name}. Reserve tickets before your visit."
                )

            lines.append("")

    return "\n".join(lines)

File: travel-agent/src/agents/sites-agent.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { Type, type Static } from "@sinclair/typebox";
import type { AgentTool, AgentToolResult } from "@mariozechner/pi-agent-core";
import { attractions } from "../data/attractions.js";
import type { Attraction } from "../data/types.js";

const SitesQuerySchema = Type.Object({
  query: Type.String({
    description:
      "Search query for attractions (name, category, or 'all')",
  }),
  day_of_week: Type.Optional(
    Type.String({
      description:
        "Day to check availability (e.g., 'Monday'). Every match is still returned; the ones closed that day are flagged as unavailable.",
    })
  ),
});

type SitesQuery = Static<typeof SitesQuerySchema>;

export function createSitesAgentTool(): AgentTool<typeof SitesQuerySchema, Attraction[]> {
  return {
    name: "query_sites",
    label: "Query Sites",
    description:
      "Query information about attractions in the city including opening hours, closure days, visit duration, ticket prices, and booking requirements. Use this to look up attraction details before planning routes.",
    parameters: SitesQuerySchema,
    execute: async (toolCallId, args, signal, onUpdate) => {
      return querySites(args);
    },
  };
}

export function querySites(args: SitesQuery): AgentToolResult<Attraction[]> {
  const { query, day_of_week } = args;

  let filtered: Attraction[];

  if (query.toLowerCase() === "all") {
    filtered = [...attractions];
  } else {
    const lowerQuery = query.toLowerCase();
    filtered = attractions.filter(
      (a) =>
        a.name.toLowerCase().includes(lowerQuery) ||
        a.category.toLowerCase() === lowerQuery
    );
  }

  // day_of_week asks "is this open that day?", so answer it in place. Dropping the
  // closed attractions here would delete the very closure the route planner needs:
  // plan_route already detects a closure conflict and reschedules around it, and the
  // caller would instead be told the attraction does not exist.

  // Build text content with booking advisories
  const lines: string[] = [];

  if (filtered.length === 0) {
    lines.push(`No attractions found matching "${query}".`);
  } else {
    for (const attraction of filtered) {
      lines.push(`**${attraction.name}** (${attraction.category})`);
      lines.push(`  ${attraction.description}`);
      lines.push(
        `  Hours: ${attraction.openTime}–${attraction.closeTime}`
      );
      if (attraction.closureDays.length > 0) {
        lines.push(`  Closed: ${attraction.closureDays.join(", ")}`);
      }
      if (day_of_week && attraction.closureDays.includes(day_of_week)) {
        lines.push(
          `  ⚠️ NOT AVAILABLE on ${day_of_week}. Schedule it on another day instead.`
        );
      }
      lines.push(`  Visit duration: ${attraction.visitDurationMinutes} minutes`);
      lines.push(`  Ticket price: $${attraction.ticketPrice}`);

      if (attraction.advanceBookingRequired) {
        lines.push(
          `  ⚠️ BOOKING ADVISORY: Advance booking is required for ${attraction.name}. Reserve tickets before your visit.`
        );
      }

      lines.push("");
    }
  }

  return {
    content: [{ type: "text", text: lines.join("\n") }],
    details: filtered,
  };
}
